### DISCUSSION TITLE
Is it possible to see if a resource has a Pulumi state?

### DISCUSSION THREAD

**Discussion by bra-fsn** [pulumi/pulumi#17539] [Pulumi in Practice]
I would like to automatically (no manual imports) manage already existing AWS resources with Pulumi, as it offers an easily usable, nice API.

The idea is to get the current resources, import them into Pulumi and make any modifications from there.
For this to work, I would need to know if a resource has already been imported into Pulumi and if yes, create a different resource with the modifications.
Doing this on import is not possible, as the two resources must be the same.
I've tried to look up the Python SDK docs to see if I can query the resources under a given stack, but couldn't find it.

Take for example an AWS CloudWatch log group.

I would like to do something like this:

```python
import pulumi_aws as aws
import pulumi

loggroups = aws.cloudwatch.get_log_groups(
    log_group_name_prefix="/test",
    opts=pulumi.InvokeOptions(provider=provider),
)
for group in loggroups.log_group_names:
    lg = aws.cloudwatch.get_log_group(name=group, opts=pulumi.InvokeOptions(provider=provider))
    lg_opts = dict(
        name=group,
        kms_key_id=lg.kms_key_id,
        log_group_class=lg.log_group_class,
        retention_in_days=lg.retention_in_days,
        opts=pulumi.ResourceOptions(provider=provider, import_=group, ignore_changes=["tags", "tagsAll"]),
    )
    if already_imported():
        lg_opts |= dict(retention_in_days=14, opts=pulumi.ResourceOptions(provider=provider))
    aws.cloudwatch.LogGroup(group, **lg_opts)
```

I'm looking for something, which can provide the `already_imported()` result, so the imported resource can be changed.

**Comment by t0yv0** [comment]
This is an interesting question.

If you did not want to make modifications from Pulumi and just track references to cloud resources in state but not "own" the resources as in making changes, there is the https://www.pulumi.com/docs/iac/concepts/resources/get/ feature. I believe this will create references in the Pulumi state and allow your program read-only access to the resource information but not editing them.

If you do want to import and own the resources, there are two forms - the import resource option you are using as well as `pulumi import` command. Pulumi import command can also accept a list of resources to import in bulk. This could be useful, but it is difficult to fully automate as `pulumi import` suggests a code edit, which ideally is reviewed by a human. You could build something at Automation API level that finds resources, inspects the current state to find new resources not yet tracked, then performs bulk import on these new resources and suggests a code edit. 

I'll ask around if there are any good built-in ways to solve this with the features Pulumi has, the product is growing fast.

**Reply by bra-fsn** [reply]
My real problem is that I would like to import the existing resource and make a change without code change.
So instead of doing this:

1. import the resource
2. change the code (remove the import option, change the resource)

I would like to do it with the same, unchanged code.
I'm fine if it's not possible to do the import and resource change in one step, but it should be possible to do it by executing the code twice (first: import the resource unchanged, second: change the resource).
I'm looking for a solution to this, so either importing and changing the resource in one step or make it possible to do it in two steps (with the same code, without external state).

**Reply by t0yv0** [reply]
Hm, interesting, I am not sure what your expectations are of import here. If a new resource appears in the cloud and you import it somehow without changing Pulumi sources or auxillary data that Pulumi sources are reading, the desired state described by your Pulumi stack will not have this new resource. This means two things:

 - if this resource is in Pulumi state, subsequently `pulumi up` will attempt to delete it
 - if this resource is not in Pulumi state, `pulumi up` will ignore it, but then this resource is not "in" Pulumi in any practical way.

I think the only exception to this that I know of is the Resource Getter feature: https://www.pulumi.com/docs/iac/concepts/resources/get/ 

With this feature you can get a "read-only" handle on a resource that is kept in Pulumi state and can be inspected, but Pulumi does not manage this resource and will not attempt to modify or delete it.

**Reply by bra-fsn** [reply]
> Hm, interesting, I am not sure what your expectations are of import here. If a new resource appears in the cloud and you import it somehow without changing Pulumi sources or auxillary data that Pulumi sources are reading, the desired state described by your Pulumi stack will not have this new resource. This means two things:
> 
>     * if this resource is in Pulumi state, subsequently `pulumi up` will attempt to delete it
> 
>     * if this resource is not in Pulumi state, `pulumi up` will ignore it, but then this resource is not "in" Pulumi in any practical way.
> 
> 
> I think the only exception to this that I know of is the Resource Getter feature: https://www.pulumi.com/docs/iac/concepts/resources/get/
> 
> With this feature you can get a "read-only" handle on a resource that is kept in Pulumi state and can be inspected, but Pulumi does not manage this resource and will not attempt to modify or delete it.

I'm trying to reiterate my problem:

* our (AWS) resources are managed by Pulumi
* sometimes new resources are created by external software directly, via AWS APIs
   * I have no control over these resources as they are created
   * these resources are "write once" from the side of that external software: they are created by them and never touched
   * I want to modify these resources after they are created and because we use Pulumi extensively, it would be great to be able to use that

I give a simple and theoretical example, I hope it will make it more clear:

1. our AWS account have several S3 buckets, we use tag based cost attribution (so we label all buckets as `Owner=our_team`)
2. the external program occasionally creates new S3 buckets. Let's say it has created a bucket named `external_program_bucket_5555`
3. this bucket won't have our tags as we can't modify the way the external program manages its AWS resources, but we want to have those tags on our buckets
4. as we use Pulumi, it would be great to be able to do this:
   1. run a static Pulumi program (a Pulumi stack, calling `pulumi up`) regularly
   2. this would list all S3 buckets (with the `get_` methods) and see if there are new ones, which match our filter (let's say the bucket name starts with: `external_program_*`) and if there are matching buckets, creates Pulumi resources for them
   3. this is the point where we would need to distinguish between the "resource is already managed by Pulumi" and "not yet managed by Pulumi" state. If I could check if a given resource (even just by it's unique Pulumi resource name) already exists in the Pulumi state, I could have a branch in my program, which does this:
      1. if the resource is not yet in Pulumi's state: `aws.s3.Bucket(bucket_name, bucket=bucket_name,opts=pulumi.ResourceOptions(import_=bucket_name))` -> **imports the resource into Pulumi's state as is**
      2. if the resource already exists in Pulumi state (we've already imported it): `aws.s3.Bucket(bucket_name, bucket=bucket_name,tags={"Owner":"our_team"})` -> **this will be a Pulumi change op, adds the new tag to the already existing resource (and remove the `import_` option)**

The code would be static (no change if a new bucket appears) and regularly executed. It would be driven by the list of available S3 buckets.
If a new bucket appears and matches the filter, it imports the bucket (unchanged).
On next execution if it finds and already existing bucket, it uses the wanted bucket resource (with tags).

**Comment by t0yv0** [comment]
I've logged https://github.com/pulumi/pulumi/issues/17661 a well. This is really interesting.
